A user can edit details of a bedspace

bryangaledxw opened the pull request:
# Changes in this PR

* A user can edit the details of a bedspace

This PR begins with some minor tidying to existing integration/E2E tests

## Screenshots of UI changes

![localhost_3000_properties_f99b65ed-e2bf-46a6-81f8-a95a2c2a4d07_bedspaces_c8fe9f85-d577-4217-908f-b9cc99e4b33c](https://user-images.githubusercontent.com/94137563/200847867-7dba1021-41b1-401f-9283-207ae00483c6.png)

bryangaledxw committed:
Add update method to RoomClient

This is our API method for updating the details of a room. We also wiremock the corresponding endpoint, add stubs for integration testing, and add an UpdateRoom factory for our unit test

bryangaledxw committed:
Add update method to BedspaceService

bryangaledxw committed:
Extract shared partial _editable.njk template from our new bedspace page

As a lot of form fields will be shared between our new bedspace and edit bedspace page, we extract this partial template for later use. Will also extract a base class from our integration test helper BedspaceNewPage to mirror this

bryangaledxw committed:
Add edit request handler to our BedspacesController

This will be responsible for displaying our edit bedspace page. We also add a new method on BedspaceService to fetch a room and marshal its data into the correct format for pre-populating the form, our edit.njk template, and a stub integration test helper, BedspsaceEditPage, for later use

In our tests, we need to set the request params in a beforeEach block, to prevent state leakages between tests causing failures

bryangaledxw committed:
Link to bedspace editing page

We now route the previously created edit request handler on BedspacesController, and update our link on our bedspace show page. We also add a method to our integration test helper PremisesShowPage to use this link

bryangaledxw committed:
Add find method to RoomClient

We will use this to view get the details of a single bedspace to pre-populate the edit bedspace form. We also add the corresponding wiremocks and stubs for integration tests

bryangaledxw committed:
Add update request handler to BedspacesController

This uses the update method on BedspaceService we created earlier to do the actual work of updating our premises

bryangaledxw committed:
Add route to update request handler on BedspacesController

Now that this request handler is reachable, we update our bedspace edit form to point to it

bryangaledxw committed:
Correct naming and comment in premises integration tests

bryangaledxw committed:
Add integration tests for editing a bedspace

bryangaledxw committed:
Namespace E2E test steps

Ahead of adding E2E tests for our bedspace features, we pull out a new step definition file for our premises features. This is to avoid further growing a single test file with all "manage" related test steps

bryangaledxw committed:
In our integration tests, isolate a particular bedspace in our premises show page by name rather than room database ID

This will help us later in our E2E tests, where we won't necessarily know the database ID of a given bedspace's room

bryangaledxw committed:
Fix premises E2E tests not changing characteristics

Previously, our test was clearing a premises' characteristics, but then re-setting the same characteristics. We now set a new pair of characteristics when editing a premises

bryangaledxw committed:
Correct spelling in premises E2E tests

bryangaledxw committed:
Explicitly set a premises ID to "unknown" in our premises E2E tests

In our E2E tests, we are creating a premises to correspond to a premises in the database, without knowing what its ID will be. Rather than using the autogenerated ID from the factory, we set it to "unknown", to aid any future debugging. We will follow this convention in future E2E tests

bryangaledxw committed:
Add E2E tests for creating and editing a bedspace

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [A user can edit details of a bedspace](https://trello.com/c/0SfeOMYF/502-a-user-can-edit-details-of-a-bedspace)

richpjames commented on integration_tests/mockApis/room.ts:
```diff
@@ -54,4 +54,27 @@ export default {
+        url: paths.premises.rooms.update({ premisesId: args.premisesId, roomId: args.room.id }),
+      },
+      response: {
+        status: 201,
```
Should this be a 200?

richpjames commented on wiremock/roomStub.ts:
```diff
@@ -49,10 +55,25 @@ rooms.push({
+    urlPathPattern: `/premises/${guidRegex}/rooms/${guidRegex}`,
+  },
+  response: {
+    status: 201,
```
flagging this as well

richpjames reviewed (approved):
LGTM 
One minor thing: I think the commit message for a commit you squashed into e8cc6e945a4131da9f980eeddf44da4bbd6775a4 is still included

bryangaledxw committed:
Update generated types

We also update our StaffMember and NewArrival factories to match the updated type, and make a minor change in ArrivalsController to keep legacy Approved Premises code compiling
